File: library-test/src/androidTest/java/com/uphyca/creditcardedittext/subject/CreditCardNumberEditTextSubject.java
package com.uphyca.creditcardedittext.subject;

import com.google.common.truth.FailureMetadata;
import com.google.common.truth.Subject;
import com.uphyca.creditcardedittext.CreditCardBrand;
import com.uphyca.creditcardedittext.CreditCardNumberEditText;

import static com.google.common.truth.Truth.assertAbout;

/**
 * Subject for {@link CreditCardNumberEditText} instances.
 */
public final class CreditCardNumberEditTextSubject extends Subject {

    public static CreditCardNumberEditTextSubject assertThat(CreditCardNumberEditText editText) {
        return assertAbout(editTexts()).that(editText);
    }

    private static Subject.Factory<CreditCardNumberEditTextSubject, CreditCardNumberEditText> editTexts() {
        return CreditCardNumberEditTextSubject::new;
    }

    private final CreditCardNumberEditText actual;

    private CreditCardNumberEditTextSubject(FailureMetadata failureMetadata, CreditCardNumberEditText actual) {
        super(failureMetadata, actual);
        this.actual = actual;
    }

    public final void hasNumber(String number) {
        check("getNumber()")
                .that(actual.getNumber())
                .isEqualTo(number);
    }

    public final void hasTextString(String number) {
        check("getNumber()")
                .that(actual.getText().toString())
                .isEqualTo(number);
    }

    public final void hasBrand(CreditCardBrand brand) {
        check("getNumber()")
                .that(actual.getBrand())
                .isEqualTo(brand);
    }
}

File: library-test/src/androidTest/java/com/uphyca/creditcardedittext/subject/CreditCardNumberEditTextSubject.kt
package com.uphyca.creditcardedittext.subject

import com.google.common.truth.FailureMetadata
import com.google.common.truth.Subject
import com.google.common.truth.Subject.Factory
import com.google.common.truth.Truth.assertAbout
import com.uphyca.creditcardedittext.CreditCardBrand
import com.uphyca.creditcardedittext.CreditCardNumberEditText

/**
 * Subject for [CreditCardNumberEditText] instances.
 */
class CreditCardNumberEditTextSubject private constructor(
    failureMetadata: FailureMetadata,
    private val actual: CreditCardNumberEditText
) : Subject(failureMetadata, actual) {

    fun hasNumber(number: String?) {
        check("getNumber()")
            .that(actual.number)
            .isEqualTo(number)
    }

    fun hasTextString(number: String?) {
        check("getNumber()")
            .that(actual.text.toString())
            .isEqualTo(number)
    }

    fun hasBrand(brand: CreditCardBrand?) {
        check("getNumber()")
            .that(actual.brand)
            .isEqualTo(brand)
    }

    companion object {

        @JvmStatic
        fun assertThat(editText: CreditCardNumberEditText?): CreditCardNumberEditTextSubject {
            return assertAbout(editTexts()).that(editText)
        }

        private fun editTexts(): Factory<CreditCardNumberEditTextSubject, CreditCardNumberEditText> {
            return Factory { failureMetadata, actual ->
                CreditCardNumberEditTextSubject(failureMetadata, actual)
            }
        }
    }
}
